### DISCUSSION TITLE
Self-Hosted runners: runs-on tag with multiple labels does not work as documented

### DISCUSSION THREAD

**Discussion by Ryker-LabJack** [community/community#50172] [Actions]
### Select Topic Area

Bug

### Body

Hello,

I have been setting up github actions to run with our self-hosted runners.
The workflow is meant to run on each of our runner, and produce artifacts based on the OS and ARCH.
To delegate the builds, I setup a matrix job with OS and ARCH specified, correlating to the default tags applied to the runner.
The matrix variables are then applied to the 'runs-on' tag in the workflow to assign the job to the correct machine.
Bellow are some pictures of the workflow and self-hosted runners:

![workflow](https://user-images.githubusercontent.com/109097999/225435969-34ce468e-470b-4d9b-b412-41140d8eff15.png)
![self-hosted-labels](https://user-images.githubusercontent.com/109097999/225436038-9a4651e0-c5db-4587-9f8e-b33a6e0851b6.png)

This is where the problems start...
The runs-on tag produces a yaml error from github: `You have an error in your yaml syntax`
Which is odd because I believe I am following the instructions from the [documentation](https://docs.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow#using-custom-labels-to-route-jobs).
`To select runners that meet certain criteria, such as operating system or architecture, we recommend providing an array of labels that begins with self-hosted (this must be listed first) and then includes additional labels as needed.`
But what ever, take out the array brackets and now github will accept the workflow file, but now the jobs will be queued without ever executing.

After more trial and error, I found the only way for the workflow to accept the workflow and deploy the jobs was to have only 1 label for runs-on, even though the documentation says it can handle [multiple labels](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idruns-on)
`If you specify an array of strings, your workflow will execute on any runner that matches all of the specified runs-on values.`
Only applying one label does not exactly work for my use case:
- `self-hosted` will trigger all the runners, but applies random matrix values to each
- `${{ matrix.os }}` applies to the correct OS, but will mixup architectures
- `${{ matrix.arch }}` applies to the correct ARCH, but will mixup os's
- Another approach is to use the runner group, but this has the same results as `self-hosted`

I thought spelling/grammar/format was a problem, since the [documentation](https://docs.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow#using-default-labels-to-route-jobs) states that self-hosted runners are by default assigned an OS and ARCH label which is what I need.
But the examples from the documentation shows lowercase, while github reports capitalization, but neither formatting had any affect on multi-label use.
I also tried with quotes and without quotes in the matrix options.
The repository I am running from is private, and I have checked the repository actions settings and the organization settings that the repo is part of and where the self hosted runners are connected to make sure the proper settings and permissions are set.

I eventually found a work around for this problem by assigning a unique label to each runner, then adding that label to the matrix for each entry, and only using that matrix label for the runs-on tag

I feel like that proper solution is the one I started with and the one outlined in the documentation, but I was unable to get that to work, and I would appreciate any insight on why I was having these problems.

TLDR: runs-on does not work with multiple labels and self-hosted runners

**Comment by franke-biosaxs** [comment]
@Ryker-LabJack I do not have a solution for you, but I am as well bashing my head over the excessively helpful "You have an error in your yaml syntax" error message. My situation is the same as you describe, essentially:
`runs-on: [ self-hosted, ${{ matrix.os }} ]`
does not work as one would expect from the documentation; `[ self-hosted, explicit-os-name ]` works fine, though. 

I'm glad I found your post, as I shall use the unique-label workaround as well.

**Comment by tylermilner** [comment]
@Ryker-LabJack, @franke-biosaxs I just ran into the same `You have an error in your yaml syntax` error using `runs-on: [ self-hosted, ${{ matrix.os }} ]`. I was able to get this to work by using alternate yaml syntax for arrays (dashed `-` list) rather than `[ ]` brackets. Check it out:
```yaml
  test:
    strategy:
      matrix:
        include:
          - ios: ios-12
            os: macos-12
          - ios: ios-16
            os: macos-13
    needs: lint
    runs-on:
      - self-hosted
      - ${{ matrix.os }}
```
In my case, the `macos-12` and `macos-13` labels are custom assigned to my runners since the default `macOS` and `X64` labels didn't give me the build host OS version granularity I needed, but I would expect it to work with the default labels as well since there is no longer the yaml syntax error.

**Comment by weineran** [comment]
Try enclosing the parameterized values in quotes like this:
```
runs-on: [ self-hosted, "${{ matrix.os }}" ]
```

I was doing something similar and that worked for me.
